Deploy crunchy postgres application as k8s statefulset

nsathyaseelan opened the pull request:
Signed-off-by: sathyaseelan <sathyaseelan.n@cloudbyte.com>

<!--  Thanks for sending a pull request!  Here are some tips for you -->

**What this PR does / why we need it**:

- Deploys postgres statefulset application with 2 application replicas.
- Verifying if all the replicas are created and in running state.

@ksatchit @gprasath @dargasudarshan can you please review this PR

nsathyaseelan commented:
Limus job Logs:

```
sathyaseelan_n@cloudshell:~/sathya-pr/litmus/apps/crunchy-postgress/deployers (strong-eon-153112)$ kubectl logs litmus-pgset-slkmn -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "2d768ec560d23e1f4062fa401b9bf024a0472948", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "a638f0da192a587cb8d883e43678e567", "mode": "0644", "owner": "root", "size": 416, "src": "/root/.ansible/tmp/ansible-tmp-1535701422.46-259465817169087/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.486863", "end": "2018-08-31 07:43:44.744235", "failed_when_result": false, "rc": 0, "start": "2018-08-31 07:43:43.257372", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"postgres-deploy\" configured", "stdout_lines": ["litmusresult \"postgres-deploy\" configured"]}

TASK [Check whether the provider storageclass is applied] **********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get sc openebs-standard", "delta": "0:00:00.890710", "end": "2018-08-31 07:43:45.843665", "rc": 0, "start": "2018-08-31 07:43:44.952955", "stderr": "", "stderr_lines": [], "stdout": "NAME               PROVISIONER\nopenebs-standard   openebs.io/provisioner-iscsi", "stdout_lines": ["NAME               PROVISIONER", "openebs-standard   openebs.io/provisioner-iscsi"]}

TASK [Replace the pvc placeholder with provider] *******************************
changed: [127.0.0.1] => {"changed": true, "msg": "2 replacements made"}

TASK [Replace the storageclass placeholder with provider] **********************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Create test specific namespace.] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "source ~/.profile; kubectl create ns postgres", "delta": "0:00:00.908323", "end": "2018-08-31 07:43:47.529864", "rc": 0, "start": "2018-08-31 07:43:46.621541", "stderr": "mesg: ttyname failed: Inappropriate ioctl for device", "stderr_lines": ["mesg: ttyname failed: Inappropriate ioctl for device"], "stdout": "namespace \"postgres\" created", "stdout_lines": ["namespace \"postgres\" created"]}

TASK [Checking the status  of test specific namespace.] ************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get ns postgres -o custom-columns=:status.phase --no-headers", "delta": "0:00:00.912603", "end": "2018-08-31 07:43:48.630141", "rc": 0, "start": "2018-08-31 07:43:47.717538", "stderr": "", "stderr_lines": [], "stdout": "Active", "stdout_lines": ["Active"]}

TASK [Deploying postgres statefulset and services] **************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f postgres-statefulset.yaml -n postgres", "delta": "0:00:01.094613", "end": "2018-08-31 07:43:49.930297", "rc": 0, "start": "2018-08-31 07:43:48.835684", "stderr": "", "stderr_lines": [], "stdout": "serviceaccount \"pgset-sa\" created\nservice \"pgset\" created\nservice \"pgset-primary\" created\nservice \"pgset-replica\" created\nstatefulset \"pgset\" created", "stdout_lines": ["serviceaccount \"pgset-sa\" created", "service \"pgset\" created", "service \"pgset-primary\" created", "service \"pgset-replica\" created", "statefulset \"pgset\" created"]}

TASK [Verify that the postgres master and replica are available as cluster services] ***
            - name: Obtain the number of replicas.
changed: [127.0.0.1] => {"changed": true, "cmd": "source ~/.profile; kubectl get svc -n postgres", "delta": "0:00:01.113084", "end": "2018-08-31 07:43:51.304886", "failed_when_result": false,
"rc": 0, "start": "2018-08-31 07:43:50.191802", "stderr": "mesg: ttyname failed: Inappropriate ioctl for device", "stderr_lines": ["mesg: ttyname failed: Inappropriate ioctl for device"], "stdout": "NAME                                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE\npgset                                         ClusterIP   None
  <none>        5432/TCP            2s\npgset-primary                                 ClusterIP   10.63.252.191   <none>        5432/TCP            2s\npgset-replica
      ClusterIP   10.63.253.40    <none>        5432/TCP            2s\npostgres-pgdata-pgset-0-3904356226-ctrl-svc   ClusterIP   10.63.254.37    <none>        3260/TCP,9501/TCP   1s", "stdout_lines": ["NAME                                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE", "pgset                                         ClusterIP   None
        <none>        5432/TCP            2s", "pgset-primary                                 ClusterIP   10.63.252.191   <none>        5432/TCP            2s", "pgset-replica
                ClusterIP   10.63.253.40    <none>        5432/TCP            2s", "postgres-pgdata-pgset-0-3904356226-ctrl-svc   ClusterIP   10.63.254.37    <none>        3260/TCP,9501/TCP
1s"]}

TASK [Obtain the number of replicas.] ******************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get sts -n postgres -o custom-columns=:spec.replicas", "delta": "0:00:01.006958", "end": "2018-08-31 07:43:52.627609",
"rc": 0, "start": "2018-08-31 07:43:51.620651", "stderr": "", "stderr_lines": [], "stdout": "\n2", "stdout_lines": ["", "2"]}

TASK [Obtain the ready replica count and compare with the replica count.] ******
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (15 retries left).
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (14 retries left).
changed: [127.0.0.1] => {"attempts": 3, "changed": true, "cmd": "kubectl get sts -n postgres -o custom-columns=:..readyReplicas", "delta": "0:00:00.868023", "end": "2018-08-31 07:45:55.847109", "rc": 0, "start": "2018-08-31 07:45:54.979086", "stderr": "", "stderr_lines": [], "stdout": "\n2", "stdout_lines": ["", "2"]}

TASK [Checking postgress replica pods are running] **********************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n postgres -o jsonpath='{.items[?(@.metadata.labels.app==\"pgset\")].status.phase}'", "delta": "0:00:00.922668", "end": "2018-08-31 07:45:56.979739", "rc": 0, "start": "2018-08-31 07:45:56.057071", "stderr": "", "stderr_lines": [], "stdout": "Running Running", "stdout_lines": ["Running Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "05e339c40d4c436a27c41be22bdee5372dfd4449", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "c047e2c833bdf5e91d9447811bca11f5", "mode": "0644", "owner": "root", "size": 414, "src": "/root/.ansible/tmp/ansible-tmp-1535701557.13-275680735959991/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.010267", "end": "2018-08-31 07:46:00.136215", "failed_when_result": false, "rc": 0, "start": "2018-08-31 07:45:59.125948", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"postgres-deploy\" configured", "stdout_lines": ["litmusresult \"postgres-deploy\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=16   changed=14   unreachable=0    failed=0
```

```
sathyaseelan_n@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n litmus -a
NAME                 READY     STATUS      RESTARTS   AGE
litmus-pgset-9tl4w   0/1       Completed   0          4m

sathyaseelan_n@cloudshell:~ (strong-eon-153112)$ kubectl get pods -n postgres
NAME                                                       READY     STATUS    RESTARTS   AGE
pgset-0                                                    1/1       Running   0          3m
pgset-1                                                    1/1       Running   0          2m
postgres-pgdata-pgset-0-1008560481-ctrl-668fbf7446-6qmc4   2/2       Running   0          3m
postgres-pgdata-pgset-0-1008560481-rep-d7b79b4ff-5ffpt     1/1       Running   0          3m
postgres-pgdata-pgset-0-1008560481-rep-d7b79b4ff-jwbn2     1/1       Running   0          3m
postgres-pgdata-pgset-0-1008560481-rep-d7b79b4ff-z9hnr     1/1       Running   0          3m
postgres-pgdata-pgset-1-374197817-ctrl-769bf75d9c-9t95l    2/2       Running   0          2m
postgres-pgdata-pgset-1-374197817-rep-75ff7f88f7-6kqv9     1/1       Running   0          2m
postgres-pgdata-pgset-1-374197817-rep-75ff7f88f7-8pgmk     1/1       Running   0          2m
postgres-pgdata-pgset-1-374197817-rep-75ff7f88f7-d5f9g     1/1       Running   0          2m
```

gprasath commented on apps/crunchy-postgres/deployers/postgres-installer.yaml:
```diff
@@ -0,0 +1,133 @@
+          shell: kubectl apply -f {{ pg_statefulset }} -n {{ namespace }}
+          args:
+            executable: /bin/bash
+          become: true
```
Want to be root here?

gprasath commented on apps/crunchy-postgres/deployers/postgres-installer.yaml:
```diff
@@ -0,0 +1,133 @@
+          register: result
+
+        - name: Verify that the postgres master and replica are available as cluster services
+          shell: source ~/.profile; kubectl get svc -n {{ namespace }}
```
remove profile

ksatchit commented on apps/crunchy-postgres/deployers/deploy-postgres-litmus.yaml:
```diff
@@ -0,0 +1,36 @@
+            value: openebs-standard
+
+          - name: PROVIDER_PVC
+            value: pgdata
```
Can we follow a convention here ? pgdata-claim

ksatchit commented on apps/crunchy-postgres/deployers/postgres-installer.yaml:
```diff
@@ -0,0 +1,132 @@
+
+        - block:
+
+            - name: Obtain the number of replicas.
```
Can we use app labels here (cluster may contain multiple sts)

ksatchit commented on apps/crunchy-postgres/deployers/postgres-statefulset.yaml:
```diff
@@ -0,0 +1,111 @@
+      storageClassName: testclass
+      resources:
+        requests:
+          storage: 5G
```
:+1:

nsathyaseelan commented:
Sample output after modification:-
```
sathyaseelan_n@cloudshell:~/sathya-pr/litmus/apps/crunchy-postgres/deployers (strong-eon-153112)$ kubectl logs -f litmus-pgset-992gb -n litmus
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "2d768ec560d23e1f4062fa401b9bf024a0472948", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "a638f0da192a587cb8d883e43678e567", "mode": "0644", "owner": "root", "size": 416, "src": "/root/.ansible/tmp/ansible-tmp-1535781251.31-206240624288617/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.310342", "end": "2018-09-01 05:54:13.468871", "failed_when_result": false, "rc": 0, "start": "2018-09-01 05:54:12.158529", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"postgres-deploy\" configured", "stdout_lines": ["litmusresult \"postgres-deploy\" configured"]}

TASK [Check whether the provider storageclass is applied] **********************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get sc openebs-standard", "delta": "0:00:00.949715", "end": "2018-09-01 05:54:14.640441", "rc": 0, "start": "2018-09-01 05:54:13.690726", "stderr": "", "stderr_lines": [], "stdout": "NAME               PROVISIONER\nopenebs-standard   openebs.io/provisioner-iscsi", "stdout_lines": ["NAME               PROVISIONER", "openebs-standard   openebs.io/provisioner-iscsi"]}

TASK [Replace the pvc placeholder with provider] *******************************
changed: [127.0.0.1] => {"changed": true, "msg": "2 replacements made"}

TASK [Replace the storageclass placeholder with provider] **********************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Create test specific namespace.] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "source ~/.profile; kubectl create ns postgres", "delta": "0:00:00.946619", "end": "2018-09-01 05:54:16.387785", "rc": 0, "start": "2018-09-01 05:54:15.441166", "stderr": "mesg: ttyname failed: Inappropriate ioctl for device", "stderr_lines": ["mesg: ttyname failed: Inappropriate ioctl for device"], "stdout": "namespace \"postgres\" created", "stdout_lines": ["namespace \"postgres\" created"]}

TASK [Checking the status  of test specific namespace.] ************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get ns postgres -o custom-columns=:status.phase --no-headers", "delta": "0:00:00.943808", "end": "2018-09-01 05:54:17.530132", "rc": 0, "start": "2018-09-01 05:54:16.586324", "stderr": "", "stderr_lines": [], "stdout": "Active", "stdout_lines": ["Active"]}

TASK [Deploying postgres statefulset and services] *****************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f postgres-statefulset.yaml -n postgres", "delta": "0:00:01.843193", "end": "2018-09-01 05:54:19.592919", "rc": 0, "start": "2018-09-01 05:54:17.749726", "stderr": "", "stderr_lines": [], "stdout": "serviceaccount \"pgset-sa\" created\nservice \"pgset\" created\nservice \"pgset-primary\" created\nservice \"pgset-replica\" created\nstatefulset \"pgset\" created", "stdout_lines": ["serviceaccount \"pgset-sa\" created", "service \"pgset\" created", "service \"pgset-primary\" created", "service \"pgset-replica\" created", "statefulset \"pgset\" created"]}

TASK [Verify that the postgres master and replica are available as cluster services] ***
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get svc -n postgres", "delta": "0:00:01.048499", "end": "2018-09-01 05:54:20.852919", "failed_when_result": false, "rc": 0, "start": "2018-09-01 05:54:19.804420", "stderr": "", "stderr_lines": [], "stdout": "NAME                                              TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE\npgset                                             ClusterIP   None            <none>        5432/TCP            2s\npgset-primary                                     ClusterIP   10.63.252.26    <none>        5432/TCP            2s\npgset-replica                                     ClusterIP   10.63.241.76    <none>        5432/TCP            1s\npostgres-pgdata-claim-pgset-0-73728498-ctrl-svc   ClusterIP   10.63.254.230   <none>        3260/TCP,9501/TCP   1s", "stdout_lines": ["NAME                                              TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE", "pgset                                             ClusterIP   None            <none>        5432/TCP            2s", "pgset-primary                                     ClusterIP   10.63.252.26    <none>        5432/TCP            2s", "pgset-replica                                     ClusterIP   10.63.241.76    <none>        5432/TCP            1s", "postgres-pgdata-claim-pgset-0-73728498-ctrl-svc   ClusterIP   10.63.254.230   <none>        3260/TCP,9501/TCP   1s"]}

TASK [Obtain the number of replicas.] ******************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get sts -n postgres -o jsonpath='{.items[?(@.metadata.labels.app==\"pgset\")].spec.replicas}'", "delta": "0:00:01.043620", "end": "2018-09-01 05:54:22.175180", "rc": 0, "start": "2018-09-01 05:54:21.131560", "stderr": "", "stderr_lines": [], "stdout": "2", "stdout_lines": ["2"]}

TASK [Obtain the ready replica count and compare with the replica count.] ******
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (15 retries left).
FAILED - RETRYING: Obtain the ready replica count and compare with the replica count. (14 retries left).
changed: [127.0.0.1] => {"attempts": 3, "changed": true, "cmd": "kubectl get sts -n postgres -o jsonpath='{.items[?(@.metadata.labels.app==\"pgset\")].status.readyReplicas}'", "delta": "0:00:00.917949", "end": "2018-09-01 05:56:25.722453", "rc": 0, "start": "2018-09-01 05:56:24.804504", "stderr": "", "stderr_lines": [], "stdout": "2", "stdout_lines": ["2"]}

TASK [Checking postgress replicas pods are running state] **********************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n postgres -o jsonpath='{.items[?(@.metadata.labels.app==\"pgset\")].status.phase}'", "delta": "0:00:00.987367", "end": "2018-09-01 05:56:26.922236", "rc": 0, "start": "2018-09-01 05:56:25.934869", "stderr": "", "stderr_lines": [], "stdout": "Running Running", "stdout_lines": ["Running Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "05e339c40d4c436a27c41be22bdee5372dfd4449", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "c047e2c833bdf5e91d9447811bca11f5", "mode": "0644", "owner": "root", "size": 414, "src": "/root/.ansible/tmp/ansible-tmp-1535781387.07-58736501252952/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.056588", "end": "2018-09-01 05:56:30.206321", "failed_when_result": false, "rc": 0, "start": "2018-09-01 05:56:29.149733", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"postgres-deploy\" configured", "stdout_lines": ["litmusresult \"postgres-deploy\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=16   changed=14   unreachable=0    failed=0
```

ksatchit commented:
good one @nsathyaseelan !
